Generate installcheck-good schedules from autoconf settings

danielgustafsson opened the pull request:
As posted to gpdb-dev@ earlier:

In between sorting out merge conflicts from upstream I had a look at making `installcheck-good` be set up based on the options passed to `./configure` and the dependencies per each testsuite.  This was surprisingly trickier to get right than anticipated (as is so often the case) but I think I have a proposal to
discuss.  The hard part is not the code itself but to keep the schedule files, and the dependency management in them, easy to work with and maintain the readability of the files.

The schedule files are turned into .in files which are processed at make all into schedules with the supported testsuites kept as `test:` rows and the ones with unresolved dependencies moved to `ignore:` rows. Dependencies are set per test by suffixing the testname with a list in parenthesis:
```
test: aaa bbb(foo bar) ccc(foo)
```
Here, test aaa has no dependencies, bbb depends on `--with-foo` and `--with-bar` and ccc depends on `--with-foo`.  The exact names of the dependencies aren’t limited to the featurename as they are set in autoconf but it’s probably a good idea to maintain the name.  In the example above the following sequence would not cause bbb to be marked as FAILED as it would simply not be executed at all:
```
./configure --with-foo
make install
make -C gpAux/gpdemo cluster
make installcheck-good
```
I first implemented this in autoconf with M4sh but it became quite unreadable so I started over using Perl which is the language used for most tools in `src/test/regress`.  I was hoping to avoid introducing more bespoke tools but writing this in M4sh or make would become less maintainable.  The result is a
new small tool called [gensched.pl](https://github.com/danielgustafsson/gpdb/blob/configure_icg/src/test/regress/gensched.pl) which takes the features set in autoconf as parameters, reads the .in file on STDIN and outputs the schedule on STDOUT (invocation can be seen in the regress GNUMakefile).

With something like this in place we obviously need to ensure that tests only depend on their core features and not (too many) auxilliary ones which would cause tests to be skipped (which today would fail instead which arguably isn’t better).

danielgustafsson commented:
(this code was branched off before the gpfdist code was pushed, thus the merge conflict)

hlinnaka commented:
Probably better to keep parallel_schedule unchanged, as that comes from PostgreSQL.

ascherbaum-pivotal commented on src/test/regress/gensched.pl:
```diff
@@ -0,0 +1,84 @@
+				foreach my $d (@deps)
+				{
+					push(@ignore, $t) && next T_LOOP
+						if (!defined($dependencies{$d}));
```
Having the if at the end of the statement is not helpful reading this statement. I know, it's Perl, still ...

danielgustafsson replied:
Well, it's language X written in the style of language X.. but it can be rewritten.

Looking at it though I think it would've been even more readable by doing this instead of the if:
```Perl
.. unless(defined($dependencies{$d}));
```

danielgustafsson commented:
Rebased on top of master and force pushed.

andreasscherbaum commented:
This rebased patch reduces the number of errors in my build from 7 to 5.
All the mapreduce errors are gone.

ascherbaum-pivotal commented:
@hlinnaka Except the "parallel_schedule", any other objections?

This patch will reduce a number regression test problems and I would like to see that it is merged.

hlinnaka commented:
On 21/03/16 23:36, Andreas Scherbaum wrote:
> @hlinnaka Except the "parallel_schedule", any other objections?

no objections.

- Heikki

danielgustafsson commented:
I will revisit this wrt the parallel_schedule comment within short.

ascherbaum-pivotal commented:
@danielgustafsson Any update on this patch?
If you rebase it and fix the parallel_schedule, I can test it for you.

danielgustafsson commented:
> On 05 Apr 2016, at 13:32, Andreas Scherbaum <notifications@github.com> wrote:
> 
> @danielgustafsson Any update on this patch?
> If you rebase it and fix the parallel_schedule, I can test it for you.

Thanks! It’s been put on the backburner a bit with the merge but I’ll try to get some time to fix it up.

danielgustafsson commented:
I revisited this and addressed the comments raised in this PR as well as rebased on top of a current master. While there I also completed the greenplum_schedule file with all the dependencies we have right now.

danielgustafsson commented:
The longer this lingers the less I'm convinced this is worth the added complexity. Will close unless there are objections.

danielgustafsson commented:
Closing due to inactivity.

Eulerizeit commented:
@pivotal-mike This clearly won't work but it seems like what you are looking for. We need to make sure the tests are in the correct place.
